### ISSUE TITLE
when switching from stream8 to rl8 ipa fails

### ISSUE THREAD

**Issue by RobVerduijn** [rocky-linux/rocky-tools#75]
Hello,

When you are running an ipa server on stream8 the migrate script will downgrade the ipa packages.
A downgraded ipa refuses to start.

Reverting back to stream8 fixes the problem.

Maybe include an option in the script to exclude a certain package set ?

Rob

**Comment by msz59** [comment]
Does it mean that (apart from problems such as with IPA) migrate script can be run to convert CentOS Stream installation to RL?

**Comment by pajamian** [comment]
Migration from stream 8 isn't officially supported.  That said, it should work for the most part, but you stumbled across one potential issue.  The problem is that migration does a distro-sync and since stream 8 packages are newer than those in 8 the sync will cause those packages to downgrade.  In some cases (as you have discovered), the config files, binary data formats, etc of older versions is not forward-compatible, so this downgrade causes the programs to not be able to read, parse, or otherwise understand these files which were generated by the newer version.

The real solution here is to document this as a known issue in the README file.

**Comment by msz59** [comment]
Thanks for the explanation.
Out of curiosity - when all that buzz about dumping CentOS 8 by RedHat started and new distros were announced to become C8.x/RHEL8.x binary replacement (as Rocky or Alma), I remember people saying that the migration would be just a matter of replacing some "metadata", like xxxx-release and xxx-repos, not even requiring rebooting the machine. Now from what I see, your script, even when migrating an up-to-date CentOS 8.4 system to RL 8.4, is physically reinstalling all packages which takes quite a while. Is this really the only way to do it?

**Comment by RobVerduijn** [comment]
> Thanks for the explanation.
> Out of curiosity - when all that buzz about dumping CentOS 8 by RedHat started and new distros were announced to become C8.x/RHEL8.x binary replacement (as Rocky or Alma), I remember people saying that the migration would be just a matter of replacing some "metadata", like xxxx-release and xxx-repos, not even requiring rebooting the machine. Now from what I see, your script, even when migrating an up-to-date CentOS 8.4 system to RL 8.4, is physically reinstalling all packages which takes quite a while. Is this really the only way to do it?

It will work I guess.
Or rather it's what I did when I reverted back to centos stream8.
Replaced 3 rpms manually and then ran yum update to get ipa back up to date.

be carefull if you try this,  it will try to eat your face
ensure you have a backup and you practiced how to restore it
Rob

ps the 3 rpms: 
- centos-gpg-keys-8-2.el8.noarch.rpm
- centos-stream-release-8.5-3.el8.noarch.rpm
- centos-stream-repos-8-2.el8.noarch.rpm

**Comment by pajamian** [comment]
migrate2rocky attempts to do a migration that will work for many different possible scenarios on different source distributions, so it, by necessity entails a bit more than just replacing a few metaddata packages.  You may be able to get away with just replacing those few packages on your particular system, though.

Just replacing the metadata has been known to work on the odd occasion, but a proper migration requires a distro-sync to make sure that you're on the same packages offered by the new (target) distribution.

The three packages mentioned are a subset, there may be more depending on what is initially installed in the source distribution.

**Comment by RobVerduijn** [comment]
@pajamian totally agree with you, better to do a proper migration  than just wing it and see what happens.
however when things go pear shaped you have to get creative.

I usually prefer reinstalling systems but that's not always an option.
Rob

**Comment by pajamian** [comment]
I'm currently working on other code to help recover a system when things go bad with a migration.  That in addition to documenting this particular issue may help.  As I said, stream migrations are not strictly supported, but I will try to document how to recover a system if someone tries it and has issues, the main thing I think is to find the stream packages in question and update them again from their CentOS Stream 8 repositories, then just wait for RL8 to catch up in the next point release.  This will necessarily have to be a manual post-migration step as there is no way for migrate2rocky to detect this type of failure (in fact, as far as migrate2rocky is concerned the migration was a success).

**Comment by pajamian** [comment]
@RobVerduijn Can you tell me exactly which IPA packages (and from which module(s)) were causing this issue?  Then I can possibly come up with some steps to fix it post-migration manually and document those.  I will try to make it more generic because this issue certainly applies to other packages (not just IPA) but I can use your case as an example.

**Comment by RobVerduijn** [comment]
@pajamian that will take some testing and time.
I'll setup a test environment and see what I can dig up.

**Comment by pajamian** [comment]
It occurs to me that it may be possible to safely migrate a CentOS stream box, if we leave the stream repos and packages in place, but disable the stream repos then the migration would result in a mixed RockyLinux / CentOS Stream system which should, in theory, update to all RockyLinux packages once the next point release of RockyLinux drops.

**Comment by RobVerduijn** [comment]
just found out the hard way that uefi will bite you hard if you only replace 'some metadata'
took some effort to recover from that.
reinstalling grub2-efi-x64 and shim-x64 AFTER  you switch and BEFORE you reboot is a wise move

**Comment by pajamian** [comment]
> just found out the hard way that uefi will bite you hard if you only replace 'some metadata'
> took some effort to recover from that.
> reinstalling grub2-efi-x64 and shim-x64 AFTER you switch and BEFORE you reboot is a wise move
> 
> maybe even throw in the following for good measure
> efibootmgr -c -d /dev/<disk> -p 1 -l \EFI\rocky\shimx64.efi -L "Rocky Linux 8"
> or if you are switching back
> efibootmgr -c -d /dev/<disk> -p 1 -l \EFI\centos\shimx64.efi -L "Centos 8"

Right, migrate2rocky takes these steps when completing a migration on an EFI boot system.  It is, as I said, just one more thing that complicates migrations to make them more than just replacing a few system packages.

**Comment by RobVerduijn** [comment]
Hi,

Sorry for the delay, 
took a while to find time to do this

I first had a nice routine that checked which package was the latest, after which I found out that if you do that there are conflicts that prevent you from updateing. The --nobest option is required in that case.

Anyway the below script downloads the packages from a centos8 stream mirror and then installs them.

```
#!/bin/bash
  
URL=http://mirror.wd6.net/centos/8-stream/AppStream/x86_64/os/Packages
RPMDIR=/root/rpms
PUBDNS="8.8.8.8"
RPMS=''

# ensure your dns is working
cp /etc/resolv.conf{,.bak}
echo "nameserver ${PUBDNS}" > /etc/resolv.conf

# create rpm list
RPMS="${RPMS} 389-ds-base-1.4.3.23-2.module_el8.5.0+835+5d54734c.x86_64.rpm"
RPMS="${RPMS} 389-ds-base-legacy-tools-1.4.3.23-2.module_el8.5.0+835+5d54734c.x86_64.rpm"
RPMS="${RPMS} 389-ds-base-libs-1.4.3.23-2.module_el8.5.0+835+5d54734c.x86_64.rpm"
RPMS="${RPMS} ipa-client-4.9.3-1.module_el8.5.0+836+f9e1ecc9.x86_64.rpm"
RPMS="${RPMS} ipa-client-common-4.9.3-1.module_el8.5.0+836+f9e1ecc9.noarch.rpm"
RPMS="${RPMS} ipa-common-4.9.3-1.module_el8.5.0+836+f9e1ecc9.noarch.rpm"
RPMS="${RPMS} ipa-server-4.9.3-1.module_el8.5.0+836+f9e1ecc9.x86_64.rpm"
RPMS="${RPMS} ipa-server-common-4.9.3-1.module_el8.5.0+836+f9e1ecc9.noarch.rpm"
RPMS="${RPMS} ipa-server-dns-4.9.3-1.module_el8.5.0+836+f9e1ecc9.noarch.rpm"
RPMS="${RPMS} ipa-server-trust-ad-4.9.3-1.module_el8.5.0+836+f9e1ecc9.x86_64.rpm"
RPMS="${RPMS} python3-ipaclient-4.9.3-1.module_el8.5.0+836+f9e1ecc9.noarch.rpm"
RPMS="${RPMS} python3-ipalib-4.9.3-1.module_el8.5.0+836+f9e1ecc9.noarch.rpm"
RPMS="${RPMS} python3-ipaserver-4.9.3-1.module_el8.5.0+836+f9e1ecc9.noarch.rpm"
RPMS="${RPMS} python3-lib389-1.4.3.23-2.module_el8.5.0+835+5d54734c.noarch.rpm"

# create download dir
[ ! -e ${RPMDIR} ] && mkdir -p ${RPMDIR}

pushd ${RPMDIR}

# download rpms
for RPM in ${RPMS}; do
  if [ ! -e ${RPMDIR}/${RPM} ] ; then
    curl -OL ${URL}/${RPM}
  fi  
done

dnf update -y $(ls *.rpm)

popd

# restore dns
cp /etc/resolv.conf{.bak,}
rm -f /etc/resolv.conf.bak
```

if for some reason you want the latest and greatest
the next script prints out a list  that you can copy and paste into the previous script
(or you source it into the previous one, whatever makes you happy)
```
#!/bin/bash
  
URL=http://mirror.wd6.net/centos/8-stream/AppStream/x86_64/os/Packages
PACKAGES=''
RPMS=''

PACKAGES="${PACKAGES} 389-ds-base"
PACKAGES="${PACKAGES} 389-ds-base-legacy-tools"
PACKAGES="${PACKAGES} 389-ds-base-libs"
PACKAGES="${PACKAGES} ipa-client"
PACKAGES="${PACKAGES} ipa-client-common"
PACKAGES="${PACKAGES} ipa-common"
PACKAGES="${PACKAGES} ipa-server"
PACKAGES="${PACKAGES} ipa-server-common"
PACKAGES="${PACKAGES} ipa-server-dns"
PACKAGES="${PACKAGES} ipa-server-trust-ad"
PACKAGES="${PACKAGES} python3-ipaclient"
PACKAGES="${PACKAGES} python3-ipalib"
PACKAGES="${PACKAGES} python3-ipaserver"
PACKAGES="${PACKAGES} python3-lib389"

echo 'RPMS=""'
export RPMS=''
for PACKAGE in ${PACKAGES}; do
  RPM=$(curl -s ${URL}/ | sed -n "s/.*\(${PACKAGE}-[0-9].*rpm\).*/\1/p" | sort | tail -n1)
  echo 'RPMS="${RPMS} '${RPM}'"'
  export RPMS="${RPMS} $RPM"
done
```

cheers
Rob

**Comment by pajamian** [comment]
@RobVerduijn Can you test the migration with this version of the script, please:
https://github.com/pajamian/rocky-tools/blob/stream8/migrate2rocky/migrate2rocky.sh

**Comment by pajamian** [comment]
Fixed with #96
